crypto_kem/newhope1024cca/clean/poly.c

```diff
@@ -0,0 +1,370 @@
+
+        r[7 * i + 0] =  t0 & 0xff;
+        r[7 * i + 1] = (unsigned char) ((t0 >> 8) | (t1 << 6));
+        r[7 * i + 2] = (unsigned char) ((t1 >> 2));
+        r[7 * i + 3] = (unsigned char) ((t1 >> 10) | (t2 << 4));
+        r[7 * i + 4] = (unsigned char) ((t2 >> 4));
+        r[7 * i + 5] = (unsigned char) ((t2 >> 12) | (t3 << 2));
+        r[7 * i + 6] = (unsigned char) ((t3 >> 6));
+    }
+}
+
+/*************************************************
+* Name:        poly_compress
+*
+* Description: Compression and subsequent serialization of a polynomial
+*
+* Arguments:   - unsigned char *r: pointer to output byte array
+*              - const poly *p:    pointer to input polynomial
+**************************************************/
+void PQCLEAN_NEWHOPE1024CCA_CLEAN_poly_compress(unsigned char *r, const poly *p) {
+    unsigned int i, j, k = 0;
+
+    uint32_t t[8];
+
+    for (i = 0; i < NEWHOPE_N; i += 8) {
+        for (j = 0; j < 8; j++) {
+            t[j] = coeff_freeze(p->coeffs[i + j]);
+            t[j] = (((t[j] << 3) + NEWHOPE_Q / 2) / NEWHOPE_Q) & 0x7;
+        }
+
+        r[k]   =  (unsigned char) (t[0]       | (t[1] << 3) | (t[2] << 6));
+        r[k + 1] = (unsigned char) ((t[2] >> 2) | (t[3] << 1) | (t[4] << 4) | (t[5] << 7));
+        r[k + 2] = (unsigned char) ((t[5] >> 1) | (t[6] << 2) | (t[7] << 5));
+        k += 3;
+    }
+}
+
+/*************************************************
+* Name:        poly_decompress
+*
+* Description: De-serialization and subsequent decompression of a polynomial;
+*              approximate inverse of poly_compress
+*
+* Arguments:   - poly *r:                pointer to output polynomial
+*              - const unsigned char *a: pointer to input byte array
+**************************************************/
+void PQCLEAN_NEWHOPE1024CCA_CLEAN_poly_decompress(poly *r, const unsigned char *a) {
+    unsigned int i, j;
+    for (i = 0; i < NEWHOPE_N; i += 8) {
+        r->coeffs[i + 0] =  a[0] & 7;
+        r->coeffs[i + 1] = (a[0] >> 3) & 7;
+        r->coeffs[i + 2] = (a[0] >> 6) | ((a[1] << 2) & 4);
+        r->coeffs[i + 3] = (a[1] >> 1) & 7;
+        r->coeffs[i + 4] = (a[1] >> 4) & 7;
+        r->coeffs[i + 5] = (a[1] >> 7) | ((a[2] << 1) & 6);
+        r->coeffs[i + 6] = (a[2] >> 2) & 7;
+        r->coeffs[i + 7] = (a[2] >> 5);
+        a += 3;
+        for (j = 0; j < 8; j++) {
+            r->coeffs[i + j] = ((uint32_t)r->coeffs[i + j] * NEWHOPE_Q + 4) >> 3;
+        }
+    }
+}
+
+/*************************************************
+* Name:        poly_frommsg
+*
+* Description: Convert 32-byte message to polynomial
+*
+* Arguments:   - poly *r:                  pointer to output polynomial
+*              - const unsigned char *msg: pointer to input message
+**************************************************/
+void PQCLEAN_NEWHOPE1024CCA_CLEAN_poly_frommsg(poly *r, const unsigned char *msg) {
+    unsigned int i, j, mask;
+    for (i = 0; i < 32; i++) { // XXX: MACRO for 32
```

Comment:
Should this XXX be resolved?

Reply:
This is also in https://github.com/newhopecrypto/newhope/blob/master/ref/poly.c 
@cryptojedi?

Reply:
Resolved